File: source/math/Float2.java
package math;

import java.io.Serializable;

public class Float2 implements Serializable {
    public float x, y;

    public Float2() {
        x = 0.0f;
        y = 0.0f;
    }

    public Float2(float a) {
        x = a;
        y = a;
    }

    public Float2(float[] data) {
        this(data[0], data[1]);
    }

    public Float2(float x, float y) {
        this.x = x;
        this.y = y;
    }

    public Float2(Float2 v) {
        x = v.x;
        y = v.y;
    }

    public Float2(Int2 v) {
        x = (float)v.x;
        y = (float)v.y;
    }

    // Getter
    public float[] array() {
        return new float[] { x, y };
    }

    // Addition
    public Float2 add(Float2 v) {
        return new Float2(x + v.x, y + v.y);
    }

    // Subtraction
    public Float2 sub(Float2 v) {
        return new Float2(x - v.x, y - v.y);
    }

    // Multiplication
    public Float2 mul(float a) {
        return new Float2(x * a, y * a);
    }
    public Float2 mul(Float2 v) {
        return new Float2(x * v.x, y * v.y);
    }

    // Division
    public Float2 div(float a) {
        return mul(1.0f / a);
    }
    public Float2 div(Float2 v) {
        return new Float2(x / v.x, y / v.y);
    }

    // Comparison
    public boolean equals(Float2 v) {
        return x == v.x && y == v.y;
    }

    // Negation
    public Float2 neg() {
        return mul(-1.0f);
    }

    // Absolute
    public Float2 abs() {
        return new Float2(Math.abs(x), Math.abs(y));
    }

    // Length
    public float len() {
        return (float) Math.sqrt(x * x + y * y);
    }

    // Normalization
    public Float2 norm() {
        return div(len());
    }

    // Dot product
    public float dot(Float2 v) {
        return x * v.x + y * v.y;
    }

    // Angle between vectors
    public float angle(Float2 v) {
        return (float) Math.toDegrees(Math.acos(norm().dot(v.norm())));
    }

    // Returns a rotated vector around the given point
    public Float2 rotate(float angle) {
        final float sinA = (float) Math.sin(Math.toRadians(angle));
        final float cosA = (float) Math.cos(Math.toRadians(angle));
        return new Float2(cosA * x - sinA * y, sinA * x + cosA * y);
    }

    // Constants
    public static Float2 getPosX() {
        return new Float2(1.0f, 0.0f);
    }
    public static Float2 getNegX() {
        return new Float2(-1.0f, 0.0f);
    }
    public static Float2 getPosY() {
        return new Float2(0.0f, 1.0f);
    }
    public static Float2 getNegY() {
        return new Float2(0.0f, -1.0f);
    }

    // String
    public String toString() {
        return "(" + x + ", " + y + ")";
    }
}

File: source/math/Float2.kt
package math

import java.io.Serializable
import kotlin.math.acos
import kotlin.math.cos
import kotlin.math.sin
import kotlin.math.sqrt

class Float2 : Serializable {
    var x: Float = 0f
    var y: Float = 0f

    constructor(x: Float, y: Float) {
        this.x = x
        this.y = y
    }

    constructor() : this(0f, 0f)

    constructor(a: Float) : this(a, a)

    constructor(data: FloatArray) : this(data[0], data[1])

    constructor(v: Float2) : this(v.x, v.y)

    constructor(v: Int2) : this(v.x.toFloat(), v.y.toFloat())

    // Getters
    val length: Float
        get() = sqrt(this * this)

    val array: FloatArray
        get() = floatArrayOf(x, y)

    // Math
    operator fun plus(v: Float2): Float2 {
        return Float2(x + v.x, y + v.y)
    }

    operator fun minus(v: Float2): Float2 {
        return Float2(x - v.x, y - v.y)
    }

    operator fun times(a: Float): Float2 {
        return Float2(x * a, y * a)
    }

    operator fun times(v: Float2): Float {
        return x * v.x + y * v.y
    }

    operator fun div(a: Float): Float2 {
        return this * (1f / a)
    }

    operator fun unaryMinus(): Float2 {
        return this * -1f
    }

    override fun equals(other: Any?): Boolean {
        if (other is Float2) {
            return x == other.x && y == other.y
        }
        return false
    }

    fun angle(v: Float2): Float {
        return toDegrees(acos(normalize(this) * normalize(v)))
    }

    fun rotate(angle: Float): Float2 {
        val sinA = sin(toRadians(angle))
        val cosA = cos(toRadians(angle))
        return Float2(cosA * x - sinA * y, sinA * x + cosA * y)
    }

    override fun toString(): String {
        return "($x, $y)"
    }

    override fun hashCode(): Int {
        var result = x.hashCode()
        result = 31 * result + y.hashCode()
        return result
    }

    companion object {
        val posX: Float2
            get() = Float2(1f, 0f)
        val negX: Float2
            get() = Float2(-1f, 0f)
        val posY: Float2
            get() = Float2(0f, 1f)
        val negY: Float2
            get() = Float2(0f, -1f)
    }
}